karate-core/src/main/java/com/intuit/karate/core/ScenarioEngine.java

```diff
@@ -1061,33 +1091,65 @@ private void attachVariables() {
         return detached;
     }
 
+    /**
+     * Note that if there is a possibility of the root object passed into this method being a PolyglotMap or PolyglotList
+     * a new Map/List instance will be produced thus the result of this method should be re-used.
+     *
+     * Typically this is used in scenarios where the "root" object is a host Map/List attached to the Scenario Engine (vars) or
+     * magicVariables etc.
+     *
+     * Any non-function values of the Map/List are still stored in that map/list by reference so the concept of context share
+     * from Karate is maintained.
+     *
+     * @param o
+     * @return
+     */
     protected Object recurseAndAttach(Object o) {
         if (o instanceof Value) {
             Value value = (Value) o;
             return value.canExecute() ? attach(value) : null;
+        } else if (o instanceof Function && this.JS != null) {
```

Comment:
just 2 comments:
* this routine of "transforming" objects, I would explore whether they can be done in 1`JsValue` in that case they apply to all of karate
* I'd really want to keep the new features for the sake of tests to a minimum, just from experience it is a headache to understand and maintain in the long term. I'd try to fit everything into the existing parallel package and max add 1-2 files, or if you drop a note with the "bare minimum" to replicate, I can think through a bit